Kernel/Arch/riscv64/PCI/Initializer.cpp

```diff
@@ -4,17 +4,159 @@
+    auto const& device_tree = DeviceTree::get();
+
+    auto maybe_soc = device_tree.get_child("soc"sv);
+    if (!maybe_soc.has_value()) {
+        dmesgln("PCI: No `soc` node found in the device tree, PCI initialization will be skipped");
+        return;
+    }
+
+    auto const& soc = maybe_soc.value();
+
+    enum class ControllerCompatible {
+        CAM,
+        ECAM,
+    };
+
+    // These properties must be present
+    auto soc_address_cells = soc.get_property("#address-cells"sv).value().as<u32>();
+    [[maybe_unused]] auto soc_size_cells = soc.get_property("#size-cells"sv).value().as<u32>();
+
+    Optional<u32> domain_counter;
+    for (auto const& entry : soc.children()) {
+        if (!entry.key.starts_with("pci"sv))
+            continue;
+        auto const& node = entry.value;
+
+        if (auto device_type = node.get_property("device_type"sv); !device_type.has_value() || device_type.value().as_string() != "pci"sv) {
+            dmesgln("PCI: PCI named device tree entry {} not a PCI controller, got device type {} instead", entry.key, device_type.has_value() ? device_type.value().as_string() : "<None>"sv);
+            continue;
+        }
+
+        auto maybe_compatible = node.get_property("compatible"sv);
+        if (!maybe_compatible.has_value()) {
+            dmesgln("PCI: PCI controller {} did not provide a compatible string, rejecting", entry.key);
+            continue;
+        }
+        auto compatible = maybe_compatible.value().as_strings();
+        ControllerCompatible controller_compatibility;
+        if (compatible.contains_slow("pci-host-ecam-generic"sv)) {
+            controller_compatibility = ControllerCompatible::ECAM;
+        } else if (compatible.contains_slow("pci-host-cam-generic"sv)) {
+            controller_compatibility = ControllerCompatible::CAM;
+            // FIXME: Implement CAM
+            dmesgln("PCI: PCI controller {} is CAM, which is not supported", entry.key);
+            continue;
+        } else {
+            dmesgln("PCI: PCI controller {} did not provide a compatible string, rejecting", entry.key);
+            dmesgln("PCI: Compatible strings provided: {}", compatible);
+            continue;
+        }
+
+        auto maybe_reg = node.get_property("reg"sv);
+        if (!maybe_reg.has_value()) {
+            dmesgln("PCI: PCI controller {} did not provide a physical address, rejecting", entry.key);
+            continue;
+        }
+        auto reg = maybe_reg.value();
+
+        u8 bus_range[2];
+        auto maybe_bus_range = node.get_property("bus-range"sv);
+        if (!maybe_bus_range.has_value()) {
+            bus_range[0] = 0;
+            bus_range[1] = 255;
+        } else {
+            auto bus_range = maybe_bus_range.value().as<Array<BigEndian<u32>, 2>>();
+            // FIXME: Range check
+            bus_range[0] = bus_range[0];
+            bus_range[1] = bus_range[1];
```

Comment:
You are shadowing the `bus_range` variable here, so it will only contain invalid values.